src/jdk.internal.le/share/classes/jdk/internal/org/jline/terminal/impl/ffm/CLibrary.java

```diff
@@ -0,0 +1,1250 @@
+
+    static {
+        // methods
+        Linker linker = Linker.nativeLinker();
+        SymbolLookup lookup = SymbolLookup.loaderLookup().or(linker.defaultLookup());
+        // https://man7.org/linux/man-pages/man2/ioctl.2.html
+        ioctl = linker.downcallHandle(
+                lookup.find("ioctl").get(),
+                FunctionDescriptor.of(
+                        ValueLayout.JAVA_INT, ValueLayout.JAVA_INT, ValueLayout.JAVA_LONG, ValueLayout.ADDRESS),
+                Linker.Option.firstVariadicArg(2));
+        // https://www.man7.org/linux/man-pages/man3/isatty.3.html
+        isatty = linker.downcallHandle(
+                lookup.find("isatty").get(), FunctionDescriptor.of(ValueLayout.JAVA_INT, ValueLayout.JAVA_INT));
+        // https://man7.org/linux/man-pages/man3/tcsetattr.3p.html
+        tcsetattr = linker.downcallHandle(
+                lookup.find("tcsetattr").get(),
+                FunctionDescriptor.of(
+                        ValueLayout.JAVA_INT, ValueLayout.JAVA_INT, ValueLayout.JAVA_INT, ValueLayout.ADDRESS));
+        // https://man7.org/linux/man-pages/man3/tcgetattr.3p.html
+        tcgetattr = linker.downcallHandle(
+                lookup.find("tcgetattr").get(),
+                FunctionDescriptor.of(ValueLayout.JAVA_INT, ValueLayout.JAVA_INT, ValueLayout.ADDRESS));
+        // https://man7.org/linux/man-pages/man3/ttyname.3.html
+        ttyname_r = linker.downcallHandle(
+                lookup.find("ttyname_r").get(),
+                FunctionDescriptor.of(
+                        ValueLayout.JAVA_INT, ValueLayout.JAVA_INT, ValueLayout.ADDRESS, ValueLayout.JAVA_LONG));
+        // https://man7.org/linux/man-pages/man3/openpty.3.html
+        LinkageError error = null;
+        Optional<MemorySegment> openPtyAddr = lookup.find("openpty");
+        if (openPtyAddr.isEmpty()) {
+            StringBuilder sb = new StringBuilder();
+            sb.append("Unable to find openpty native method in static libraries and unable to load the util library.");
+            List<Throwable> suppressed = new ArrayList<>();
+            try {
+                System.loadLibrary("util");
+                openPtyAddr = lookup.find("openpty");
+            } catch (Throwable t) {
+                suppressed.add(t);
+            }
+            if (openPtyAddr.isEmpty()) {
+                String libUtilPath = System.getProperty("jdk.internal.org.jline.ffm.libutil");
+                if (libUtilPath != null && !libUtilPath.isEmpty()) {
+                    try {
+                        System.load(libUtilPath);
+                        openPtyAddr = lookup.find("openpty");
+                    } catch (Throwable t) {
+                        suppressed.add(t);
+                    }
+                }
+            }
+            if (openPtyAddr.isEmpty() && OSUtils.IS_LINUX) {
+                String hwName;
+                try {
+                    Process p = Runtime.getRuntime().exec(new String[] {"uname", "-m"});
```

Comment:
This seems very much adhoc. What systems have you tested this on? 

If this is supposed to be a general method of locating a library in well-known locations on different Linux systems, maybe it needs to be elevated to a more prominent place where it can be shared by other parts of the JDK that might want to do the same. 

It seems like this is a bit of missing "glue" code from FFM -- how to actually locate a system library. I understand that it is not part of FFM proper, nor jextract, but I guess this is something that will need to be reimplemented many times, unless some common functionality is provided. We can perhaps not do that to save the world, but we can at least do that for the benefit of the JDK code, especially as I see FFM replacing JNI more and more, going forward.